integrations/tensorflow/e2e/keras/applications/CMakeLists.txt

```diff
@@ -119,3 +119,73 @@ iree_e2e_cartesian_product_test_suite(
 )
 
 ### BAZEL_TO_CMAKE_PRESERVES_ALL_CONTENT_BELOW_THIS_LINE ###
+
+iree_tf_benchmark_suite(
+  MODELS
+    MobileNetV2
+    MobileNetV3Small
+  MODEL_SCRIPT
+    applications_benchmark.py
+  MODEL_SCRIPT_ARGS
+    "--data=imagenet"
+    "--use_external_weights=True"
+  NEED_ARG_FOR_MODEL
+  BENCHMARK_MODE
+    "fp32-full-inference"
+  TARGET_BACKEND
+    iree_llvmaot
+  TARGET_ARCH
+    "CPU-ARM64-v8A"
+  COMPILATION_FLAGS
+    "--iree-llvm-target-triple=aarch64-none-linux-android29"
+    "--iree-flow-inline-constants-max-byte-length=2048"
+    "-iree-llvm-loop-unrolling=true"
+  RUNTIME_FLAGS
+    "--task_topology_group_count=3"
+)
```

Comment:
I still have concerns about the flexibility of this configuration.

For a project like IREE, I'd expect to see
* a somewhat central list of models we run benchmarks on
* a way to run all benchmarks on any available devices (phones connected over ADB, the host machine on CPU, the host machine on GPU, etc.). Something like https://github.com/apache/tvm/tree/main/apps/benchmark
* a way to correlate results from local runs with those published on a dashboard (or perhaps just feed into MLPerf like https://mlcommons.org/en/inference-edge-10/ to get comparisons against other frameworks)

I'd be more comfortable with this "benchmark suite" if it _just_ defined properties of the model. Target arch + llvm target triple should come from a separate file and apply to _all_ models. Extra compilation and runtime flags should be (ideally) removed entirely - let's make the compiler+runtime smarter if possible instead of relying on externally tuning flags for specific configurations.

Reply:
Thanks Scott for the feedback! Points taken. But I think it's hard to have those right now. Please pardon me for throwing out a wall of text in advance. :D

> I still have concerns about the flexibility of this configuration.
> 
> For a project like IREE, I'd expect to see
> 
> * a somewhat central list of models we run benchmarks on

This might be hard even in the future. I think we'll eventually have PyTorch, TFLite via TOSA etc. Not sure we can collect them in one place. Also, model benchmarks are just one category. There will be op microbenchmarks, C++ benchmarks.

I view the problem from a different angle. What we have here is just raw hooks in the source code. For users and interest parties I won't expect them to look into the source code to figure out what models we benchmark. They should be checking the perf dashboard, where they can also see concrete benchmark values. (Like we would like the normal users to read the new doc website. ;-P) Then what is needed here is very much for development for ourselves. Placing the CMake target close to the source code is preferrable here IMO.

> * a way to run all benchmarks on any available devices (phones connected over ADB, the host machine on CPU, the host machine on GPU, etc.). Something like https://github.com/apache/tvm/tree/main/apps/benchmark

Actually I added a new `run_benchmarks.py` script for doing that on Android phones. (It was added before you reviewed the first time so maybe you didn't notice it.) It probes the phone and launches suitable benchmarks and collects data back. For other platforms we can have them eventually as they are not a priority .

> * a way to correlate results from local runs with those published on a dashboard (or perhaps just feed into MLPerf like https://mlcommons.org/en/inference-edge-10/ to get comparisons against other frameworks)

I think that's a nice-to-have feature in the long run. IMO, that places more focus on benchmarking **phones** to see how each phone behave under IREE. It's better to have that once IREE is more stable and can cover more architectures. Right now I think we need infrastructure focusing more on benchmarking IREE itself. So the scalability should focus more on adding more benchmarks easily. (We do want to scale among more phones for sure. But won't like the models.)

> 
> I'd be more comfortable with this "benchmark suite" if it _just_ defined properties of the model. Target arch + llvm target triple should come from a separate file and apply to _all_ models. Extra compilation and runtime flags should be (ideally) removed entirely - let's make the compiler+runtime smarter if possible instead of relying on externally tuning flags for specific configurations.

I think we have the classical trade-off between long-term directions vs. short-term needs here again. :)

In the long run, I basically agree with what you said. But right now we are just not there yet; or even somewhat quite far away on some aspects. 

Generally command-line options fall into the following categories: 1) runtime flags, 2) compilation flags for the target architecture), and 3) compilation for feature controls. 

To kill our usage of command-line options, we need to make sure there are no special runtime modes (e.g., CPU 1-thread vs. 3-thread, etc.) (for category 1), have flat binary containing multiple architectures (for category 2), and just make the default generating optimized code for all architectures (for category 3). If we achieve these, I'd say we are largely done wit building the compiler stack. ;-P I'm not sure how many additional quarters that will take. (We are certainly working toward that though!) I agree that command-line options aren't good but IMO they are just tools in the toolbox that help us with the current development stage.

Because we are just in this current stage where we have irregular support for different models among different target backends, unfortunately the target specification needs to be coupled with the model. (Otherwise we need to filter out incompatible combinations, which actually brings more complexity into the system.) I'm hoping this to evolve though as we make progress on all fronts. But in the meanwhile, we still need to scale across more benchmarks.

Reply:
> Actually I added a new run_benchmarks.py script for doing that on Android phones. (It was added before you reviewed the first time so maybe you didn't notice it.)

Oo that looks useful! I did miss it because I've only taken a high level look at this PR. IMO this is already well into "write a design document first" territory, with 60+ comments and 5 reviewers. My feedback is focused on what I would hope for our end state, independent from the specific implementation here. The main thing I want to see is modularity across input formats (TF SavedModels, TF Python, TFLite flatbuffers, PyTorch, ONNX models, etc.) and across targets (e.g. GPU on a user-provided phone, CPU on a farm of build machines, etc.). Joining the model, target configuration, and extra compilation flags together in the build system does not offer that flexibility.

> In the long run, I basically agree with what you said. But right now we are just not there yet; or even somewhat quite far away on some aspects.

That's why I'm pushing harder to get us to make that jump. I'd like to see us move towards a setup where performance tuning has more clear guiding principles and documentation. For example,https://github.com/google/iree/pull/5639 that set a flag for one specific model following a performance regression. Evaluating new optimizations for subsets of our tested models is fine, but the follow-up actions aren't clear enough to me. Should users be setting that flag also? What types of models / workloads / devices benefit from it? Is that flag expected to stay, or will it be made a default? If we sprinkle flags across many different files in the build system then it becomes easy to lose track of all those details.

Reply:
Oh sorry missed your reply here. :)

> IMO this is already well into "write a design document first" territory. 

Sure, happy to put up a design doc for this to describe the whole picture and where everything connects.

> The main thing I want to see is modularity across input formats (TF SavedModels, TF Python, TFLite flatbuffers, PyTorch, ONNX models, etc.) and across targets (e.g. GPU on a user-provided phone, CPU on a farm of build machines, etc.). Joining the model, target configuration, and extra compilation flags together in the build system does not offer that flexibility.

Yeah, there is a matrix here: source models x target hardware. Right now we basically specify all combinations one by one in benchmarking (as we [have](https://github.com/google/iree/blob/b2fc28b6ca7f5156c576475d8bfe5d5d26d885d6/build_tools/mako/configuration.py#L221) hardcoded phone configuration list for each model).  I'm detaching it at the target architecture dimension to make it more scalable w.r.t. more phones: you can have any phone using ARM CPU and Mali Valhall GPU now; no need to register the specific phone in some central list. It's one step towards detachment. 

> That's why I'm pushing harder to get us to make that jump. 

Agreed and I think we are certainly working towards that. Mahesh and I were just chatting about having a list of various cl options and different failing configurations and start fixing them. But again, a matter of how soon we can reach there. ;-P 

>  What types of models / workloads / devices benefit from it? Is that flag expected to stay, or will it be made a default? 

Actually this speaks the need of scaling out the model benchmarks. ;-P Without a broad set of benchmarks, it's quite easy that we just focus on a few and neglect the impact over others. I think having the benchmark scaled first that would be quite helpful to avoid overfitting (as we can see certain options are just good for some but bad for others then we won't introduce; or if it makes everything better then we should strike to make it the default).